### PR TITLE
Ddf inc (WIP)

### PR DESCRIPTION
**Pull request by bennahugo**
Work in progress

This provides the boiler plate for providing a DDF predict.

TODO:
 - mask facets to ensure model components are not double counted
 - debugging

### CODE REVIEW

**Comment by bennahugo** [comment]
@o-smirnov @JSKenyon so the degridder is plumbed in now. I have a normalization error (probably because the convolution filters are not normalized correctly, but I should be able to get to the bottom of that soon. There is also an off by a few error in the primary facet centre and there are no masks as of yet (so the prediction only works for single facet case. I will debug the centring issue and maskes tomorrow, but I think we're on the home stretch here! 

I've added an option not to apply flags (similar to CASA's option of not applying gainflags in applycal) in order to predict the model into MODEL_DATA for imaging.

Please review the changes so far. If there is anything in the structure you don't agree with now is the time to shout out loud.

File: cubical/data_handler/__init__.py [hunk 1]
```diff
@@ -37,6 +37,17 @@ def import_montblanc():
     except:
         return None, sys.exc_info()
 
+# Try to import montblanc: if not successful, remember error for later.
```
Comment (JSKenyon) [inline cubical/data_handler/__init__.py]: Update comments here.

**Comment by bennahugo** [comment]
Alright this is getting to that point where we might just be better off reimplementing the degridder ourselves. I've spent the better part of a week just trying to debug the ddfacet aspect of this with no end in sight. A reimplementation is probably going to be faster than trying to figure out where in the convolution filter generation process this is going belly up.

As far as I can tell I've followed exactly the same procedure as facet machine in initializing the degridder (you instantiate a grid machine with wmax set and visibility frequencies given) with an image size that is already padded to EstimateNpix. The gridder than initializes its convolution filters based on these parameters and they are saved to a shared dictionary which effectively caches them. However w-projection does not work, and it is not a simple conjugation of the UVW coordinates:

image
The faceting by itself limits the w effect correctly (4x4 facets), 4 in this quadrant of the image alone:
image

sources closest to the facet centre is essentially not smeared.

In the limit of n facets the effect of w(n-1) is removed.

@landmanbester can we sit together and debug the preliminary numba degridder in africanus with your KB AA filters? I have working C++ code for the Perley polyhedron faceting scheme which can be made coplanar through the Sault projection algorithm. So we can have traditional AIPS style faceting.

**Comment by sjperkins** [comment]
Note that the initial implementation of the codex faceting gridder uses the
exact same filters as DDFacet (see the PR and the spheroidal test cases).

**Comment by bennahugo** [comment]
@sjperkins this is about as good as its going to get with @cyriltasse's degridder. There is still a 2% normalization error that we could potentially get rid of if the degridder is normalized correctly by the sum of convolution weights, but I think this is a good entry point comparison for building our own numba degridder that is parallelized by row. In the low facet number regime in which we operate the DDFacet degridder is suboptimal - in many instances the dE regions only need a very small number of facets. 

I've had a chat with @o-smirnov and we've agreed that in the interm we do not have to spend any time parallelizing this code because our long term strategy should be to replace the degridder in its entirety, with a stateless thread safe version. For that we can use a classic AIPS-style faceting scheme (aka. Cornwell & Perley http://articles.adsabs.harvard.edu//full/1992A%26A...261..353C/0000353.000.html).

I'm going to apply the final few touches to the make the overlapping facets case a bit faster and then have a go with this on some real dE limited data that the current lsm approach cannot correct for.

**Comment by sjperkins** [comment]
> I've had a chat with @o-smirnov and we've agreed that in the interm we do not have to spend any time parallelizing this code because our long term strategy should be to replace the degridder in its entirety, with a stateless thread safe version. For that we can use a classic AIPS-style faceting scheme (aka. Cornwell & Perley http://articles.adsabs.harvard.edu//full/1992A%26A...261..353C/0000353.000.html).

Cool, can you briefly summarize the difference between AIPS faceting and DDFacet for me?

**Comment by bennahugo** [comment]
> > I've had a chat with @o-smirnov and we've agreed that in the interm we do not have to spend any time parallelizing this code because our long term strategy should be to replace the degridder in its entirety, with a stateless thread safe version. For that we can use a classic AIPS-style faceting scheme (aka. Cornwell & Perley http://articles.adsabs.harvard.edu//full/1992A%26A...261..353C/0000353.000.html).
> 
> Cool, can you briefly summarize the difference between AIPS faceting and DDFacet for me?

@sjperkins no w-projection. You tilt the uvw coordinates with a rotation matrix such that w points towards the new phase centre. See https://github.com/ratt-ru/bullseye/blob/master/bullseye_mo/cpu_gpu_common/baseline_transform_policies.h#L71 and https://github.com/ratt-ru/bullseye/blob/master/bullseye_mo/cpu_algorithm/templated_gridder.h#L87

**Comment by bennahugo** [comment]
The gridding/degridding then is done with anti-aliasing filters only

**Comment by bennahugo** [comment]
@sjperkins I will test this on TriA tomorrow. If it improves the dE calibration we can merge this sucker and then start working on incorporating the jones matricies.

**Comment by bennahugo** [comment]
Btw @sjperkins @landmanbester  I've made a few fixes to the phase steering code and the demonstrator in bullseye if you want to cross compare and see working faceting code in action. It should install with pip install ./bullseye (thereafter running bullseye_gui.py).  We should discuss how to proceed - i'm quite tempted to build a degridder on top of this code - essentially it is just reshuffling the templated gridder and coming up with a gather convolution policy, along with a few reversed transformations. Currently it does w -projection, but we can change this to a polyhedron scheme by changing
https://github.com/ratt-ru/bullseye/blob/master/bullseye_mo/cpu_algorithm/wrapper.cpp#L250 to use
https://github.com/ratt-ru/bullseye/blob/master/bullseye_mo/cpu_gpu_common/baseline_transform_policies.h#L68

**Comment by bennahugo** [comment]
Ok I think I have this working - here is a simple catdagger + cubical test run. This selects components from a dicomodel (corresponding to image on left) according to regions, splits each region into facets, predicts and then I image the result with DDFacet:
image

You can now specify model list as follows:
```
mycomponents.DicoModel+-mycomponents.DicoModel@deregions.reg:mycomponents.DicoModel@deregions.reg
```
with degridding parameters as follows
```
 - 15:44:43 - main               [0.2/0.2 2.7/2.7 7.8Gb] [degridding] Options for the degridder. Only in use when predicting from DicoModels using DDFacet
 - 15:44:43 - main               [0.2/0.2 2.7/2.7 7.8Gb]  - OverS ............................................. = 11
 - 15:44:43 - main               [0.2/0.2 2.7/2.7 7.8Gb]  - Support ........................................... = 7
 - 15:44:43 - main               [0.2/0.2 2.7/2.7 7.8Gb]  - Nw ................................................ = 1000
 - 15:44:43 - main               [0.2/0.2 2.7/2.7 7.8Gb]  - wmax .............................................. = 0.0
 - 15:44:43 - main               [0.2/0.2 2.7/2.7 7.8Gb]  - Padding ........................................... = 1.7
 - 15:44:43 - main               [0.2/0.2 2.7/2.7 7.8Gb]  - NDegridBand ....................................... = 16
 - 15:44:43 - main               [0.2/0.2 2.7/2.7 7.8Gb]  - MaxFacetSize ...................................... = 0.15
 - 15:44:43 - main               [0.2/0.2 2.7/2.7 7.8Gb]  - MinNFacetPerAxis .................................. = 1
```

The first splits the dicomodel into facets bound to 0.15 degrees (see the degridding section of the parset) the +- switch subtracts the 4 regions above from the model data provided by that dicomodel and then ```:...``` specifier specifies four direction dependent directions:
```
 - 15:56:43 - DDFacetSim         [1.0/2.4 4.2/5.5 7.8Gb] Initialization sequence of source provider 'output/deep2subset.DicoModel' completed
 - 15:56:43 - data_handler       [1.0/2.4 4.2/5.5 7.8Gb] Using 1 model(s) for 5 directions(s)
 - 15:56:43 - data_handler       [1.0/2.4 4.2/5.5 7.8Gb]   model 0 (weight WEIGHT):
 - 15:56:43 - data_handler       [1.0/2.4 4.2/5.5 7.8Gb]     direction 0: output/deep2subset.DicoModel-DDE_REG2(output/deep2subset.DicoModel)-DDE_REG3(output/deep2subset.DicoModel)-DDE_REG1(output/deep2subset.DicoModel)-DDE_REG4(output/deep2subset.DicoModel)
 - 15:56:43 - data_handler       [1.0/2.4 4.2/5.5 7.8Gb]     direction 1: DDE_REG1(output/deep2subset.DicoModel)
 - 15:56:43 - data_handler       [1.0/2.4 4.2/5.5 7.8Gb]     direction 2: DDE_REG2(output/deep2subset.DicoModel)
 - 15:56:43 - data_handler       [1.0/2.4 4.2/5.5 7.8Gb]     direction 3: DDE_REG3(output/deep2subset.DicoModel)
 - 15:56:43 - data_handler       [1.0/2.4 4.2/5.5 7.8Gb]     direction 4: DDE_REG4(output/deep2subset.DicoModel)

```

You could also have overlapping regions with different dE's in which case flux may be contributing to two different directions. The flux is uniformly split in such a case although we can possibly dream up more complex situations.

**Comment by bennahugo** [comment]
Alright getting sensible models out for the complicated case as discribed above for a small dataset. I'll start a DD run on A1300 tonight and see what pops out in the morning

**Comment by bennahugo** [comment]
retest this please

**Comment by bennahugo** [comment]
Alright here is the narrowband C config 3C147 data without beams:
Before:
image
Using LSM:
image

Using bolted in ddfacet degridder:
image

So slightly worse, but still a substantial improvement for a first pass @IanHeywood @KshitijT @SpheMakh @o-smirnov 

Please have a go with your MeerKAT datasets. Feedback welcome

**Comment by bennahugo** [comment]
Alright I'm happy that this works as advertised. Now we need to regenerate the reference data and possibly put in more stringent testing @o-smirnov tagging you because you are more familiar with the test infrastructure than I am.

**Pull request merged**